### PR TITLE
rework __lookup_pgfrag_room

### PR DESCRIPTION
**Pull request by milabs**

### CODE REVIEW

File: tempesta_fw/ss_skb.c [hunk 1]
```diff
@@ -92,72 +93,93 @@ ss_skb_alloc_pages(size_t len)
...
- * @return pointer to the last fragment in the page.
+ * Find a page that has room for @len bytes.
+ * Return true if the room is found, or false otherwise.
+ * If the room is found, fill @f_out that describes the location of the room.
+ *
+ * The assumption is that there can be sufficient room available at the
+ * start or at the end of a page that holds this SKB's paged fragments.
+ *
+ * For a group of paged fragments located in the same page the available
+ * headroom and tailroom are calculated. If @len is less than both of
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: "...If `@len` is **more** than both of..."?

Comment (milabs) [inline tempesta_fw/ss_skb.c]: yep

File: tempesta_fw/ss_skb.c [hunk 2]
```diff
@@ -92,72 +93,93 @@ ss_skb_alloc_pages(size_t len)
...
-			return frag;
+		for (k = i; k < skb_shinfo(skb)->nr_frags; k++) {
+			if (map[k])
+				continue;
+
+			f_this = &skb_shinfo(skb)->frags[k];
+			if (compound_head(skb_frag_page(f_this)) != p_base)
+				continue;
+
+			if ((map[k] = 1, map[i]) > 1)
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: `map[k]` and `map[i]` are different values, so assignment to first one and getting value of the second one isn't a good idea. Please don't love `,` operator so much.

Comment (milabs) [inline tempesta_fw/ss_skb.c]: It's shortcut for:
```
map[k] = 1;
if (map[i]) > 1)
    continue;
```

File: tempesta_fw/ss_skb.c [hunk 3]
```diff
@@ -92,72 +93,93 @@ ss_skb_alloc_pages(size_t len)
...
+				continue;
+
+			h_room = min(h_room, f_this->page_offset);
+			t_room = min(t_room, p_size - ss_skb_frag_len(f_this));
+			map[i] = map[i] + !!(len > h_room && len > t_room);
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: `map[i] += !!(len > h_room && len > t_room);`?

File: tempesta_fw/ss_skb.c [hunk 4]
```diff
@@ -92,72 +93,93 @@ ss_skb_alloc_pages(size_t len)
...
+			if (map[k])
+				continue;
+
+			f_this = &skb_shinfo(skb)->frags[k];
+			if (compound_head(skb_frag_page(f_this)) != p_base)
+				continue;
+
+			if ((map[k] = 1, map[i]) > 1)
+				continue;
+
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: The first iteration of the loop always goes here and it doesn't need to pass the 3 conditions, so it's better just to initialize `map[i]` before the loop.

File: tempesta_fw/ss_skb.c (hunk 2 above)
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: I see, but the operations have little to do with each other.

**Comment by krizhanovsky** [comment]
Good to merge with small cleanups and improvements

**Pull request merged**